scrapy/extensions/postprocessing.py

```diff
@@ -0,0 +1,140 @@
+"""
+Extension for processing data before they are exported to feeds.
+"""
+from io import IOBase
+from bz2 import BZ2File
+from gzip import GzipFile
+from lzma import LZMAFile
+
+from scrapy.utils.misc import load_object
+from zope.interface import Interface, implementer
+
+
+class PostProcessorPlugin(Interface):
+    """
+    Interface for plugin which will be used by PostProcessingManager
+    for pre-export data processing.
+    """
+
+    def __init__(self, file, feed_options):
+        """
+        Initialize plugin with target file to which post-processed
+        data will be written and the feed-specific options (see :setting:`FEEDS`).
+        """
+
+    def write(self, data):
+        """
+        Write `data` to plugin's target file.
+        :param data: data passed to be written to file
+        :type data: bytes
+        :return: returns number of bytes written
+        :rtype: int
+        """
+
+    def close(self):
+        """
+        Close this plugin.
+        """
+
+
+@implementer(PostProcessorPlugin)
+class GzipPlugin:
+
+    def __init__(self, file, feed_options):
+        self.file = file
+        self.feed_options = feed_options
+        compress_level = self.feed_options.get("gzip_compresslevel", 9)
+        self.gzipfile = GzipFile(fileobj=self.file, mode="wb", compresslevel=compress_level)
+
+    def write(self, data):
+        return self.gzipfile.write(data)
+
+    def close(self):
+        self.gzipfile.close()
+        self.file.close()
+
+
+@implementer(PostProcessorPlugin)
+class Bz2Plugin:
+
+    def __init__(self, file, feed_options):
+        self.file = file
+        self.feed_options = feed_options
+        compress_level = self.feed_options.get("bz2_compresslevel", 9)
+        self.bz2file = BZ2File(filename=self.file, mode="wb", compresslevel=compress_level)
+
+    def write(self, data):
+        return self.bz2file.write(data)
+
+    def close(self):
+        self.bz2file.close()
+        self.file.close()
+
+
+@implementer(PostProcessorPlugin)
+class LZMAPlugin:
+
+    def __init__(self, file, feed_options):
+        self.file = file
+        self.feed_options = feed_options
+
+        format = self.feed_options.get("lzma_format")
+        check = self.feed_options.get("lzma_check", -1)
+        preset = self.feed_options.get("lzma_preset")
+        filters = self.feed_options.get("lzma_filter")
+        self.lzmafile = LZMAFile(filename=self.file, mode="wb", format=format,
+                                 check=check, preset=preset, filters=filters)
+
+    def write(self, data):
+        return self.lzmafile.write(data)
+
+    def close(self):
+        self.lzmafile.close()
+        self.file.close()
+
+
+class PostProcessingManager(IOBase):
```

Comment:
`CsvItemExporter` actually wrapped the storage with `io.TextIOWrapper`, and `io.TextIOWrapper` expects to have a file like object having at least the attributes and methods of `io.IOBase`. So there was some error when it only got an object with `write` and `close` methods. Subclassing `io.IOBase` fixed this in addition to declaring this object as writable with the help of `writable` method.

Reply:
Nice. I did not realize this change only needed to affect the manager and not the plugins. This is definitely the cleanest solution, great job!

I would include, however, a comment above the class definition explaining why we inherit from `IOBase` here, for future references. Indicate that it’s so that exporters can treat the result of the post-processing as a file-like writable object, and you could mention the example of CsvItemExporter using TextIOWrapper.

We should also make sure, when updating the documentation, to also review the documentation of exporters and make sure that it is clearly stated what the API of the file-like object exporters get is: that of an open, writable, non-seekable IOBase object without a `fileno` property.